src/MobileDevice/widget/MobileDevice.ts

```diff
@@ -0,0 +1,151 @@
+        if (window.device) {
+            if (this.deviceReadyEvent) {
+                this.deviceReadyEvent.remove();
+                this.deviceReadyEvent = null;
+            }
+        }
+    }
+
+    private setUpWidgetDom() {
+        domConstruct.create("div", {
+            class: "mobile-device-widget",
+            id: "mobile-device"
+        }, document.body, "first");
+    }
+
+    private onDeviceReady() {
+        if (cordova.hasOwnProperty("getAppVersion")) {
+            cordova.getAppVersion.getVersionNumber((versionNumber: string) => {
+                if (versionNumber) {
+                    cordova.getAppVersion.getVersionCode((versionCode: string) => {
+                        if (versionCode) {
+                            this.mxObject.set(this.deviceIdAttribute, window.device.uuid);
+                            this.mxObject.set(this.deviceTypeAttribute, window.device.platform);
+                            this.mxObject.set(this.appVersionBuildAttribute, versionCode);
+                            this.mxObject.set(this.appVersionVersionAttribute, versionNumber);
+                            this.commit(this.mxObject);
+                        }
+                    });
+                }
+            });
+        } else {
+            window.mx.ui.error("Mobile device widget: add cordova-plugin-app-version to your project config");
+        }
+    }
+
+    private validateProps(): string {
+        let errorMessage = "";
+        if (this.onDeviceReadyAction === "callMicroflow" && !this.microflow) {
+            errorMessage = "on click microflow is required in the 'Events' tab, 'Microflow' property";
+        } else if (this.onDeviceReadyAction === "showPage" && !this.page) {
+            errorMessage = "on click page is required in the 'Events' tab, 'Page' property";
+        }
+
+        return errorMessage && `Error in mobile device widget configuration: ${errorMessage}`;
+    }
+
+    private commit(mxObject: mendix.lib.MxObject) {
+        window.mx.data.commit({
+            callback: () => {
+                if (this.onNavigateBack) {
+                    mx.ui.back();
+                }
+                this.executeMicroFlow(mxObject);
+            },
+            error: (error) => window.mx.ui.error(`Error while saving device information: ${error.message}`),
+            mxobj: mxObject
+        });
+    }
+
+    private executeMicroFlow(mxObject: mendix.lib.MxObject) {
+        const context = this.mxcontext;
+        context.setContext(mxObject.getEntity(), mxObject.getGuid());
+        if (!mxObject || !mxObject.getGuid()) {
+            return;
+        }
+        if (this.onDeviceReadyAction === "callMicroflow" && this.microflow) {
+            const getError = (error: Error) => ` ${this.microflow}: ${error.message}`;
+            window.mx.ui.action(this.microflow, {
+                context,
+                error: error => window.mx.ui.error(`Error while executing microflow ${getError(error)}`),
```

Comment:
extraction getError to a function that is only used once does not make much sense.